### PR TITLE
Forest proof and serialization changes

### PR DESCRIPTION
**Pull request by fnothaft**

### CODE REVIEW

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala [hunk 1]
```diff
@@ -47,15 +47,51 @@ trait Interval[T <: Interval[T]] extends Comparable[T] {
   def width: Long = end - start
 
   /**
-   * Determines whether Interval overlaps with other Interval T
+   * Determines whether Interval intersects with another Interval T.
+   *
+   * @param interval Another interval to compare against.
    * @return Boolean whether or not Intervals overlap
```
Comment (heuermh) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala:53]:
```
Return true if this interval overlaps with the specified interval.
...
@return True if this interval overlaps with the specified interval
```

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala [hunk 2]
```diff
@@ -47,15 +47,51 @@ trait Interval[T <: Interval[T]] extends Comparable[T] {
...
+   * @param interval Another interval to compare against.
+   * @return True if two intervals cover intersecting indices in a coordinate
+   *   space.
+   *
+   * @see overlaps
+   */
+  def covers(interval: T): Boolean
+
+  /**
+   * Compares the distance between this Interval and other Interval T.
```
Comment (heuermh) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala:83]: Compares the distance between this interval and the specified interval.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala [hunk 3]
```diff
@@ -47,15 +47,51 @@ trait Interval[T <: Interval[T]] extends Comparable[T] {
...
+   * coordinate space, two objects may cover the same range on a chromosome, but
+   * may have opposite strandedness.
+   *
+   * Essentially, this is relaxed variant of overlaps where we have projected
+   * a complex coordinate space down. If two intervals overlap, they must cover
+   * each other. However, if two interlaps cover each other, they may not
+   * overlap.
+   *
+   * @param interval Another interval to compare against.
+   * @return True if two intervals cover intersecting indices in a coordinate
```
Comment (heuermh) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala:75]:
```
Return true if the range of positions covered by this interval
intersects the range of positions the specified interval covers.
...
Essentially, this is a relaxed variant of overlaps where we have projected
a complex coordinate space down [to what?].
...
@return True if the range of positions covered by this interval
intersects the range of positions the specified interval covers
```

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala [hunk 4]
```diff
@@ -47,15 +47,51 @@ trait Interval[T <: Interval[T]] extends Comparable[T] {
...
+   * @return Greater than/equal to/less than comparison.
    */
   def compareTo(interval: T): Int
 
+  /**
+   * Provides a distance, if defined, between two Intervals.
+   *
+   * @param interval Another interval to compare against.
+   * @return Absolute value istance between two Intervals, if defined.
+   */
```
Comment (heuermh) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala:95]:
```
Provides an absolute [valued?] distance, if defined, between this interval and
the specified interval.
...
@return Absolute [valued] distance between this interval and the specified
interval, or None if not defined
```

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 5]
```diff
@@ -190,11 +254,87 @@ class RangeSearchableArray[K <: Interval[K], T: ClassTag](arr: Array[(K, T)], so
...
+    // 2. expanding backwards: here, our sort invariant helps us, but only
+    //    slightly. we backtrack in the array until we find an interval whose
+    //    distance to the current interval is greater than the largest interval
+    //    in the dataset, or undefined. if the distance is undefined, that means
+    //    we have tracked back to a coordinate plane that sorts before our
+    //    coordinate plane. if the distance is defined and greater than the
+    //    longest interval length, this implies that the end of the array
+    //    element is more than the maximum interval length away from the start
+    //    of the query region, and thus, there cannot be any intervals earlier
+    //    in the array that overlap the query region.
```
Comment (heuermh) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala:331]: good

**Review by heuermh — COMMENTED** [review]
doc suggestions

**Comment by jpdna** [comment]
This max interval method seems like it would work, though I would imagine may not perform well in the case of some track that includes  massive multi-megabase features.  At some later date we might want to consider again why the interval tree is not a good option - if we could solve whatever implementation/performance issue existed before.

**Comment by fnothaft** [comment]
> This max interval method seems like it would work, though I would imagine may not perform well in the case of some track that includes massive multi-megabase features. At some later date we might want to consider again why the interval tree is not a good option - if we could solve whatever implementation/performance issue existed before.

Yeah, although I'd assert that you probably see a general tradeoff between feature size and track density. E.g., if you're running a join on variants, you have many variants, but they're very small, whereas, if you're running a join on large features (e.g., assembly gaps?) the features are huge but you don't have many of them. I agree though, definitely something we should look at and characterize more.

**Pull request merged**